### PR TITLE
add DepthScore

### LINKED ISSUE #854: Add `DepthScore`
## 🚀 Feature

Add `DepthScore` 

### Sources:
- Paper: [A Pseudo-Metric between Probability Distributions based on Depth-Trimmed Regions
](https://arxiv.org/pdf/2103.12711.pdf)
- [Repo](https://github.com/PierreColombo/nlg_eval_via_simi_measures/)

### Motivation

The recent NLG metrics are more often based on BERT (or related) embeddings. As such, I believe, we should also start adding such metrics into `TorchMetrics` with an extra dependency on `transformers` if a user wants to use any of these metrics. The `DepthScore` metric is from a family of untrained metrics (i.e. the model is not fine-tuned on any specific task) so it should be easier for us to begin with it.

### Abstract:
The design of a metric between probability distributions is a longstanding problem motivated by numerous applications in Machine Learning. Focusing on continuous probability distributions on the Euclidean space ℝ^d, we introduce a novel pseudo-metric between probability distributions by leveraging the extension of univariate quantiles to multivariate spaces. Data depth is a nonparametric statistical tool that measures the centrality of any element x∈ℝ^d with respect to (w.r.t.) a probability distribution or a data set. It is a natural median-oriented extension of the cumulative distribution function (cdf) to the multivariate case. Thus, its upper-level sets -- the depth-trimmed regions -- give rise to a definition of multivariate quantiles. The new pseudo-metric relies on the average of the Hausdorff distance between the depth-based quantile regions w.r.t. each distribution. Its good behavior w.r.t. major transformation groups, as well as its ability to factor out translations, are depicted. Robustness, an appealing feature of this pseudo-metric, is studied through the finite sample breakdown point. Moreover, we propose an efficient approximation method with linear time complexity w.r.t. the size of the data set and its dimension. The quality of this approximation as well as the performance of the proposed approach are illustrated in numerical experiments.

### PR DESCRIPTION
**Pull request by Sohaib-Ahmed21**
## What does this PR do?
Adds the `DepthScore` metric for evaluating text generation. The implementation follows the BERTScore architecture and logic closely, as both metrics:
- Extract contextual embeddings from transformer models
- Compare sentence representations using token-level embeddings
- Support custom models, multiple references, and various configuration options

The key difference is that DepthScore measures the distance between embedding distributions using depth-based statistical methods (integrated rank-weighted depth, Wasserstein distance, MMD, etc.) instead of token-level cosine similarity.

Fixes #854 

<details>
  <summary>Before submitting</summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [ x Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
📚 Documentation preview 📚: https://torchmetrics--3318.org.readthedocs.build/en/3318/

### CODE REVIEW

**Comment by Sohaib-Ahmed21** [comment]
Please approve the workflows.

**Comment by rittik9** [comment]
Thank you @Sohaib-Ahmed21 for this pr. Will take a look...

File: src/torchmetrics/functional/text/depth_score.py [hunk 1]
```diff
@@ -0,0 +1,721 @@
...
+        if reduction == "min":
+            out.append(chunk.min())
+        elif reduction == "max":
+            out.append(chunk.max())
+        else:
+            out.append(chunk.mean())
+    return torch.stack(out, dim=0)
+
+
+def cov_matrix(x: np.ndarray, robust: bool = False) -> np.ndarray:
```
Comment (justusschock) [inline src/torchmetrics/functional/text/depth_score.py:181]: what's the issue with `torch.cov`?

sklearn is a heavy requirement just for this.

File: src/torchmetrics/functional/text/depth_score.py [hunk 2]
```diff
@@ -0,0 +1,721 @@
...
+    indicate that the predicted sentence is closer to the reference sentence.
+
+    This implementation follows the original implementation from `DEPTH_score`_.
+
+    Args:
+        preds: Predicted sentence(s) as `str`, `Sequence[str]`, or tokenized dict
+            containing `"input_ids"` and `"attention_mask"`.
+        target: Reference sentence(s) as `str`, `Sequence[str]`, multi-reference
+            `Sequence[Sequence[str]]`, or tokenized dict containing `"input_ids"` and `"attention_mask"`.
+        model_name_or_path: Hugging Face model name/path used when `model` is not provided.
```
Comment (justusschock) [inline src/torchmetrics/functional/text/depth_score.py:511]: can we unify this with model? if `model` is a string, we use that as name or path and if it's a callable we'll use it straight away?

**Review by justusschock — COMMENTED** [review]
why a are we doing all the operations here in numpy rather than pytorch? it should be easier if we don't need to do conversions, no?
